PORTADA

# Proyecto 2. 
Tareas:
- Crear df de problemas con columnas de: Tipo de dato, numero de nulos, problema principal, solucionado S/N
- añadir alguna foto (Sharknado)

## Importamos bibliotecas

In [176]:
import pandas as pd

## Abrimos el archivo 

In [177]:
df=pd.read_csv("data/GSAF5.csv",encoding="ISO-8859-1")

## Informacion inicial del archivo

### Observacion inicial del dataset (head, info, isna().sum())

In [178]:
df.head(10)

Case Number       Date  Year        Type        Country  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked            USA   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked            USA   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked            USA   
3    2016.09.17  17-Sep-16  2016  Unprovoked      AUSTRALIA   
4    2016.09.15  16-Sep-16  2016  Unprovoked      AUSTRALIA   
5  2016.09.15.R  15-Sep-16  2016        Boat      AUSTRALIA   
6    2016.09.11  11-Sep-16  2016  Unprovoked            USA   
7    2016.09.07  07-Sep-16  2016  Unprovoked            USA   
8    2016.09.06  06-Sep-16  2016  Unprovoked  NEW CALEDONIA   
9  2016.09.05.b  05-Sep-16  2016  Unprovoked            USA   

                Area                                         Location  \
0            Florida                 New Smyrna Beach, Volusia County   
1            Florida                 New Smyrna Beach, Volusia County   
2            Florida                 New Smyrna Beach, Volusia County   
3           Victoria                                 Thirteenth Beach   
4           Victoria                                      Bells Beach   
5  Western Australia                                          Bunbury   
6            Florida                    Ponte Vedra, St. Johns County   
7             Hawaii                                     Makaha, Oahu   
8     North Province                                           Koumac   
9     South Carolina  Kingston Plantation, Myrtle Beach, Horry County   

          Activity                    Name Sex   ...          Species   \
0          Surfing                    male    M  ...               NaN   
1          Surfing          Chucky Luciano    M  ...               NaN   
2          Surfing                    male    M  ...               NaN   
3          Surfing         Rory Angiolella    M  ...               NaN   
4          Surfing                    male    M  ...         2 m shark   
5          Fishing  Occupant: Ben Stratton  NaN  ...               NaN   
6           Wading                    male    M  ...    3' to 4' shark   
7         Swimming                  female    F  ...  Tiger shark, 10?   
8     Kite surfing            David Jewell    M  ...               NaN   
9  Boogie boarding          Rylie Williams    F  ...               NaN   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   
5   West Australian, 9/15/2016      2016.09.15.R-boat.pdf   
6          News4Jax, 9/11/2016  2016.09.11-PonteVedra.pdf   
7    Hawaii News Now, 9/7/2016        2016.09.07-Oahu.pdf   
8       TVANouvelles, 9/6/2016      2016.09.06-Jewell.pdf   
9            C. Creswell, GSAF  2016.09.05.b-Williams.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   
5  http://sharkattackfile.net/spreadsheets/pdf_di...   
6  http://sharkattackfile.net/spreadsheets/pdf_di...   
7  http://sharkattackfile.net/spreadsheets/pdf_di...   
8  http://sharkattackfile.net/spreadsheets/pdf_di...   
9  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.c   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.b   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.a   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.17   
4  http://sharkattackfile.net/spreadsheets/pdf_di...

In [179]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5992 entries, 0 to 5991
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Case Number             5992 non-null   object
 1   Date                    5992 non-null   object
 2   Year                    5992 non-null   int64 
 3   Type                    5992 non-null   object
 4   Country                 5949 non-null   object
 5   Area                    5590 non-null   object
 6   Location                5496 non-null   object
 7   Activity                5465 non-null   object
 8   Name                    5792 non-null   object
 9   Sex                     5425 non-null   object
 10  Age                     3311 non-null   object
 11  Injury                  5965 non-null   object
 12  Fatal (Y/N)             5973 non-null   object
 13  Time                    2779 non-null   object
 14  Species                 3058 non-null   object
 15  Inve

In [180]:
#Valores nulos
df.isna().sum()

Case Number                  0
Date                         0
Year                         0
Type                         0
Country                     43
Area                       402
Location                   496
Activity                   527
Name                       200
Sex                        567
Age                       2681
Injury                      27
Fatal (Y/N)                 19
Time                      3213
Species                   2934
Investigator or Source      15
pdf                          0
href formula                 1
href                         3
Case Number.1                0
Case Number.2                0
original order               0
Unnamed: 22               5991
Unnamed: 23               5990
dtype: int64

### Conclusiones iniciales tras Analisis inicial:


- Se repiten columnas como case number (0,19,20): Comprobar diferencias y si son iguales borrar
- Se solapan las columnas Year y Date: Agruparla en una.
- En algunos casos los datos de la columna nombre se solapan con el sexo: Buscar solucion
- Las columnas pdf, href formula y href se solapan, intentar agrupar en una
- original order posible indice

#  Modificacion del dataset

## Localizacion de problemas y limpieza del dataset por columnas

FOTO DE LA TABLA INICIAL

### Columna: "original order"

- Justificacion: pocos valores repetidos,creo que puede ser un buen indice para posteriormente poder analizar el resto de columnas
- Problema: Valores repetidos, y discontinuidad en los numeros
- Objetivo: Lista de valores unicos y continuos
- Output: Nueva columna y posible indice

In [181]:
#ordeno la tabla por los valores original order
df.sort_values("original order",ascending=True,inplace=True)
df.head(3)

Case Number       Date  Year        Type             Country  \
5991     ND.0001  1845-1853     0  Unprovoked  CEYLON (SRI LANKA)   
5990     ND.0002  1883-1889     0  Unprovoked              PANAMA   
5989     ND.0003  1900-1905     0  Unprovoked                 USA   

                  Area                             Location  Activity  \
5991  Eastern Province  Below the English fort, Trincomalee  Swimming   
5990               NaN                 Panama Bay 8ºN, 79ºW       NaN   
5989    North Carolina                       Ocracoke Inlet  Swimming   

                       Name Sex   ... Species   \
5991                   male    M  ...      NaN   
5990        Jules Patterson    M  ...      NaN   
5989  Coast Guard personnel    M  ...      NaN   

                    Investigator or Source                             pdf  \
5991                            S.W. Baker              ND-0001-Ceylon.pdf   
5990                   The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
5989  F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   

                                           href formula  \
5991  http://sharkattackfile.net/spreadsheets/pdf_di...   
5990  http://sharkattackfile.net/spreadsheets/pdf_di...   
5989  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
5991  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0001   
5990  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0002   
5989  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0003   

     Case Number.2 original order Unnamed: 22 Unnamed: 23  
5991       ND.0001              2         NaN         NaN  
5990       ND.0002              3         NaN         NaN  
5989       ND.0003              4         NaN         NaN  

[3 rows x 24 columns]

In [182]:
#Vamos a ver que valoers faltan
df["original order"].nunique()

5988

In [183]:
#Vamos a comprobar cuantos valores se repiten 
OE_ValoresRepetidos=len(df["original order"])-df["original order"].nunique()

In [184]:
#Comprobamos QUE valores se repiten
df["original order"].value_counts().head(OE_ValoresRepetidos)

5739    2
569     2
5661    2
3847    2
Name: original order, dtype: int64

In [185]:
#Comprobamos desde el 0 al numero de casos, que valores no estan, lopaso a index ya que me resuta mas facil
df2=df.set_index("original order")
for i in range(0,len(df2)):
    try:
        df2.loc[i]
    except KeyError:
        print(i)

0
1
571
3347
5724
5881


Problema de original order: 4 numeros repetidos y en el orden de casos se han saltado varios numeros
Opciones para corregirlo:
- Opcion A: Cambios en la misma columna:
        - del 2 al 568 correcto, pasamos el segundo valor 569 al 570 
        - del 571 al 3346 correcto, del3348 al primer valor 3847 hay que restarle 1
        - del 3848 al 5660 correcto, del segundo valor del 5661 a 5723 hay que sumarle 1
        - del 5725 a 5738 correcto, del segundo valor de 5739 al 5880 hay que sumarle 1
        - despues restamos 1 a todo y los casos empezaran desde 1 siendo valores unicos
- Opcion B:
        - ponerlo en orden y crear una nueva tabla con un rango de 1 a len

ELIJO LA OPCION B

In [186]:
#cambio el indice con una lista de numeros desde el 1 hasta la longitud +1
lista=[i for i in range(1,len(df)+1)]


In [187]:
# añado la columna nueva y borro la antigua
df["OriginalOrder"]=lista
df.drop("original order",axis=1,inplace=True)

### Columna "unname 23"

 ####  analisis de los valores y conocimiento del problema

In [188]:
#Valores de Unnamed 23
set(df["Unnamed: 23"])

{'Teramo', 'change filename', nan}

In [189]:
#Cantidad de veces que salen estos valores
df["Unnamed: 23"].value_counts()

Teramo             1
change filename    1
Name: Unnamed: 23, dtype: int64

#### Conclusion y output

In [190]:
# No la considero relevante asi que decido borrarla
df.drop("Unnamed: 23",axis=1,inplace=True)

### Columna "unname 22"

#### Identificacion de valores nulos de cada columna

In [158]:
#Valores de Unnamed 22
set(df["Unnamed: 22"])

{nan, 'stopped here'}

In [171]:
#Vamos a ver cuantos valores hay
df["Unnamed: 22"].value_counts()

stopped here    1
Name: Unnamed: 22, dtype: int64

In [172]:
#Solo hay un valor "stopped here", vamos a ver el registro para ver si tiene relevancia
df3=df.set_index("Unnamed: 22")
df3.sort_values("Unnamed: 22").head(1)

Case Number       Date  Year        Type Country        Area  \
Unnamed: 22                                                                 
stopped here  2006.06.18  18-Jun-06  2006  Unprovoked  BRAZIL  Pernambuco   

                                    Location       Activity  \
Unnamed: 22                                                   
stopped here  Punta Del Chifre Beach, Olinda  Body boarding   

                                 Name Sex   ... Fatal (Y/N)   Time Species   \
Unnamed: 22                                 ...                               
stopped here  Humberto Pessoa Batista    M  ...           Y  09h00      NaN   

             Investigator or Source                     pdf  \
Unnamed: 22                                                   
stopped here     globalsurfnews.com  2006.06.18-Batista.pdf   

                                                   href formula  \
Unnamed: 22                                                       
stopped here  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                           href Case Number.1  \
Unnamed: 22                                                                     
stopped here  http://sharkattackfile.net/spreadsheets/pdf_di...    2006.06.18   

             Case Number.2 OriginalOrder  
Unnamed: 22                               
stopped here    2006.06.18          4745  

[1 rows x 22 columns]

In [173]:
#El registro esta en mitad de la tabla y no lo considero importante asi que quito toda la columna 
df.drop("Unnamed: 22", axis=1,inplace=True)

In [174]:
df

Case Number         Date  Year        Type             Country  \
5991       ND.0001    1845-1853     0  Unprovoked  CEYLON (SRI LANKA)   
5990       ND.0002    1883-1889     0  Unprovoked              PANAMA   
5989       ND.0003    1900-1905     0  Unprovoked                 USA   
5988       ND.0004  Before 1903     0  Unprovoked           AUSTRALIA   
5987       ND.0005  Before 1903     0  Unprovoked           AUSTRALIA   
...            ...          ...   ...         ...                 ...   
4       2016.09.15    16-Sep-16  2016  Unprovoked           AUSTRALIA   
3       2016.09.17    17-Sep-16  2016  Unprovoked           AUSTRALIA   
2     2016.09.18.a    18-Sep-16  2016  Unprovoked                 USA   
1     2016.09.18.b    18-Sep-16  2016  Unprovoked                 USA   
0     2016.09.18.c    18-Sep-16  2016  Unprovoked                 USA   

                   Area                             Location      Activity  \
5991   Eastern Province  Below the English fort, Trincomalee      Swimming   
5990                NaN                 Panama Bay 8ºN, 79ºW           NaN   
5989     North Carolina                       Ocracoke Inlet      Swimming   
5988  Western Australia                                  NaN  Pearl diving   
5987  Western Australia                          Roebuck Bay        Diving   
...                 ...                                  ...           ...   
4              Victoria                          Bells Beach       Surfing   
3              Victoria                     Thirteenth Beach       Surfing   
2               Florida     New Smyrna Beach, Volusia County       Surfing   
1               Florida     New Smyrna Beach, Volusia County       Surfing   
0               Florida     New Smyrna Beach, Volusia County       Surfing   

                       Name Sex   ... Fatal (Y/N)   Time   Species   \
5991                   male    M  ...           Y    NaN        NaN   
5990        Jules Patterson    M  ...           Y    NaN        NaN   
5989  Coast Guard personnel    M  ...           Y    NaN        NaN   
5988                  Ahmun    M  ...           Y    NaN        NaN   
5987                   male    M  ...           Y    NaN        NaN   
...                     ...  ...  ...         ...    ...        ...   
4                      male    M  ...           N    NaN  2 m shark   
3           Rory Angiolella    M  ...           N    NaN        NaN   
2                      male    M  ...           N  10h43        NaN   
1            Chucky Luciano    M  ...           N  11h00        NaN   
0                      male    M  ...           N  13h00        NaN   

                     Investigator or Source                             pdf  \
5991                             S.W. Baker              ND-0001-Ceylon.pdf   
5990                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
5989   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
5988  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
5987       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
...                                     ...                             ...   
4                        The Age, 9/16/2016       2016.09.16-BellsBeach.pdf   
3                        The Age, 9/18/2016       2016.09.17-Angiolella.pdf   
2               Orlando Sentinel, 9/19/2016            2016.09.18.a-NSB.pdf   
1               Orlando Sentinel, 9/19/2016        2016.09.18.b-Luciano.pdf   
0               Orlando Sentinel, 9/19/2016            2016.09.18.c-NSB.pdf   

                                           href formula  \
5991  http://sharkattackfile.net/spreadsheets/pdf_di...   
5990  http://sharkattackfile.net/spreadsheets/pdf_di...   
5989  http://sharkattackfile.net/spreadsheets/pdf_di...   
5988  http://sharkattackfile.net/spreadsheets/pdf_di...   
5987  http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                  

Hasta aqui ya he modificado las columnsa unnamed 22,23 y original order!!!

### 1. Case number
- Problemas:
    - Casos repetidos
    - Nomenglatura variable

In [8]:
df["Case Number"]

0       2016.09.18.c
1       2016.09.18.b
2       2016.09.18.a
3         2016.09.17
4         2016.09.15
            ...     
5987         ND.0005
5988         ND.0004
5989         ND.0003
5990         ND.0002
5991         ND.0001
Name: Case Number, Length: 5992, dtype: object

In [9]:
#Vemos los diferentes tipos de case number:

#No hay valores nulos pero hay valores que se repiten y el formato de numero de caso varia
print(f'{len(df["Case Number"])-len(set(df["Case Number"]))} casos se repiten')

16 casos se repiten


In [10]:
#vamos a ver que casos se repiten
numeros_repetidos=df["Case Number"].value_counts().head(16)
#cambiamos el indice a case number para saber mas de estos casos
df2=df.set_index("Case Number")

In [11]:
df2.columns

Index(['Date', 'Year', 'Type', 'Country', 'Area', 'Location', 'Activity',
       'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time', 'Species ',
       'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')